File: src/main/java/toast/client/gui/clickgui/CategoryRenderer.java
package toast.client.gui.clickgui;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.font.TextRenderer;
import org.lwjgl.glfw.GLFW;
import toast.client.ToastClient;
import toast.client.gui.clickgui.components.Description;
import toast.client.gui.clickgui.components.ModeSetting;
import toast.client.gui.clickgui.components.SliderSetting;
import toast.client.gui.clickgui.components.ToggleSetting;
import toast.client.modules.Module;
import toast.client.modules.config.Setting;

import java.util.ArrayList;
import java.util.Map;

import static toast.client.ToastClient.CONFIG_MANAGER;
import static toast.client.ToastClient.MODULE_MANAGER;
import static toast.client.gui.clickgui.ClickGuiScreen.keybindPressedCategory;
import static toast.client.utils.TwoDRenderUtils.drawTextBox;
import static toast.client.utils.TwoDRenderUtils.isMouseOverRect;

public class CategoryRenderer {
    private static final ArrayList<Slider> sliders = new ArrayList<>();
    public static TextRenderer textRenderer = MinecraftClient.getInstance().textRenderer;
    public static ClickGuiSettings settings = ClickGuiScreen.settings;
    public static ClickGuiSettings.Colors colors = settings.getColors();
    public static String categoryString;
    public static Module.Category category;
    public boolean keybindPressed = false;
    public Module keybindModule = null;
    public boolean isKeyPressed = false;
    public Description description = null;
    public boolean clickedL;
    public boolean clickedR;
    public double mouseX;
    public double mouseY;
    public boolean isCategory;

    public CategoryRenderer(int mouseX, int mouseY, Module.Category category, boolean clickedL, boolean clickedR) {
        this.isCategory = true;
        categoryString = category.toString();
        CategoryRenderer.category = category;
        this.mouseX = mouseX;
        this.mouseY = mouseY;
        this.clickedL = clickedL;
        this.clickedR = clickedR;
        int catBgColor = colors.categoryBgColor;
        if (isMouseOverRect(getMouseX(), getMouseY(), getX(), getY(), getBoxWidth(), getBoxHeight())) {
            if (isClickedR()) {
                catBgColor = colors.categoryClickColor;
                settings.getPositions(categoryString).setExpanded(!settings.getPositions(categoryString).isExpanded());
                settings.savePositions();
            } else {
                catBgColor = colors.categoryHoverBgColor;
            }
        }
        drawTextBox(getXint(), getYint(), getBoxWidth(), getBoxHeight(), colors.categoryBoxColor, colors.categoryTextColor, colors.categoryPrefixColor, catBgColor, colors.categoryPrefix, category.toString());
        if (settings.getPositions(category.toString()).isExpanded()) {
            int u = 1;
            for (Module module : MODULE_MANAGER.getModulesInCategory(category)) {
                int moduleTextColor;
                int moduleBgColor;
                if (module.getEnabled()) {
                    moduleTextColor = colors.moduleOnTextColor;
                    moduleBgColor = colors.moduleOnBgColor;
                } else {
                    moduleTextColor = colors.moduleOffTextColor;
                    moduleBgColor = colors.moduleOffBgColor;
                }
                if (isMouseOverRect(getMouseX(), getMouseY(), getX(), getYIteration(u), getBoxWidth(), getBoxHeight())) {
                    if (isClickedL()) {
                        moduleBgColor = colors.moduleClickColor;
                        module.toggle();
                        ToastClient.CONFIG_MANAGER.writeModules();
                    } else if (isClickedR()) {
                        if (settings.getPositions(categoryString).getExpandedModules().contains(module.getName())) {
                            settings.getPositions(categoryString).getExpandedModules().remove(module.getName());
                        } else {
                            settings.getPositions(categoryString).getExpandedModules().add(module.getName());
                        }
                        settings.savePositions();
                    } else {
                        moduleBgColor = colors.moduleHoverBgColor;
                    }
                    description = new Description();
                    description.desc = module.getDescription();
                    description.descPosX = (int) Math.round(getX() + getBoxWidth());
                    description.descPosY = (int) Math.round(getYIteration(u));
                    description.hasDesc = true;
                }
                drawTextBox(getXint(), (int) Math.round(getYIteration(u)), getBoxWidth(), getBoxHeight(), colors.moduleBoxColor, moduleTextColor, colors.modulePrefixColor, moduleBgColor, colors.modulePrefix, module.getName());
                u++;
                if (!module.getSettings().getSettings().isEmpty()) {
                    for (String modName : settings.getPositions(categoryString).getExpandedModules()) {
                        if (modName.equals(module.getName())) {
                            for (Map.Entry<String, Setting> settingEntry : module.getSettings().getSettings().entrySet()) {
                                Setting setting = settingEntry.getValue();
                                if (!settingEntry.getKey().equals("Visible")) {
                                    if (setting.getType() == 0) {
                                        ModeSetting.render(module, settingEntry.getKey(), colors, getX(), getMouseX(), getMouseY(), u, isClickedL());
                                        u++;
                                    } else if (setting.getType() == 1) {
                                        sliders.add(SliderSetting.render(module, setting, settingEntry.getKey(), colors, getX(), getMouseX(), getMouseY(), u, isClickedL()));
                                        u++;
                                        u++;
                                    } else if (setting.getType() == 2) {
                                        ToggleSetting.render(setting, settingEntry.getKey(), colors, getX(), getMouseX(), getMouseY(), u, isClickedL());
                                        u++;
                                    }
                                }
                            }
                            ToggleSetting.render(module.getSettings().getSetting("Visible"), "Visible", colors, getX(), getMouseX(), getMouseY(), u, isClickedL());
                            description = new Description();
                            description.desc = "Whether or not to show on the hud.";
                            description.descPosX = (int) Math.round(getX() + getBoxWidth());
                            description.descPosY = (int) Math.round(getYIteration(u));
                            description.hasDesc = true;
                            u++;
                            int keybindBgColor = colors.settingOnBgColor;
                            String keybindText;
                            if (isMouseOverRect(getMouseX(), getMouseY(), getX(), getYIteration(u), getBoxWidth(), getBoxHeight())) {
                                keybindBgColor = colors.settingHoverBgColor;
                                if (isClickedL()) {
                                    this.keybindPressed = true;
                                    this.keybindModule = module;
                                }
                            }
                            if (keybindPressedCategory != null && keybindPressedCategory.keybindModule == module) {
                                keybindBgColor = colors.settingHoverBgColor;
                                keybindText = "Keybind: ...";
                                if (!isKeyPressed) {
                                    keybindPressed = true;
                                    keybindModule = module;
                                }
                            } else {
                                int key = module.getKey();
                                if (key == GLFW.GLFW_KEY_UNKNOWN) keybindText = "Keybind: NONE";
                                else keybindText = "Keybind: " + GLFW.glfwGetKeyName(key, GLFW.glfwGetKeyScancode(key));
                            }
                            drawTextBox(getXint(), (int) Math.round(getYIteration(u)), getBoxWidth(), getBoxHeight(), colors.settingBoxColor, colors.settingOnTextColor, colors.settingPrefixColor, keybindBgColor, colors.settingPrefix, keybindText);
                            u++;
                        }
                    }
                }
            }
        }
    }

    public static double getX() {
        return settings.getPositions(categoryString).getPosX();
    }

    public void setX(double newX) {
        settings.getPositions(categoryString).setPosX(newX);
    }

    public static int getBoxWidth() {
        return ClickGuiScreen.width;
    }

    public static int getBoxHeight() {
        return ClickGuiScreen.height;
    }

    public static double getY() {
        return settings.getPositions(categoryString).getPosY();
    }

    public void setY(double newY) {
        settings.getPositions(categoryString).setPosY(newY);
    }

    public static double getYIteration(int iteration) {
        return getY() + iteration + getBoxHeight() * iteration;
    }

    public int getXint() {
        return (int) Math.round(settings.getPositions(categoryString).getPosX());
    }

    public int getYint() {
        return (int) Math.round(settings.getPositions(categoryString).getPosY());
    }

    public double getMouseX() {
        return this.mouseX;
    }

    public double getMouseY() {
        return this.mouseY;
    }

    public boolean isClickedL() {
        return clickedL;
    }

    public boolean isClickedR() {
        return clickedR;
    }

    public boolean isMouseOverCat() {
        return isMouseOverRect(getMouseX(), getMouseY(), getX() - 4, getY() - 4, getBoxWidth() + 4, getBoxHeight() + 2);
    }

    public void setKeyPressed(int keyPressed) {
        if (this.keybindPressed) {
            if (keyPressed == -1) return;
            else if (keyPressed == GLFW.GLFW_KEY_BACKSPACE || keyPressed == GLFW.GLFW_KEY_DELETE)
                this.keybindModule.setKey(-1);
            else this.keybindModule.setKey(keyPressed);
            this.isKeyPressed = true;
            keybindPressedCategory = null;
            CONFIG_MANAGER.writeKeyBinds();
        } else this.keybindPressed = true;
    }

    public void updateMousePos(double mouseX, double mouseY) {
        this.mouseX = mouseX;
        this.mouseY = mouseY;
    }

    public boolean updatePosition(double dragX, double dragY) {
        if (this.isMouseOverCat() && isCategory) {
            this.setX(getX() + dragX);
            this.setY(getY() + dragY);
            settings.savePositions();
            return true;
        } else {
            if (sliders.size() > 0) {
                for (Slider slider : sliders) {
                    if (isMouseOverRect(mouseX, mouseY, slider.sliderKnobX - 4, slider.sliderPosY - 4, 10, 10)) {
                        double newSliderKnobX = mouseX;
                        if (newSliderKnobX >= slider.sliderPosX && newSliderKnobX <= slider.sliderPosX + slider.sliderBarLength) {
                            slider.sliderKnobX = newSliderKnobX;
                            slider.setting.setValue((((mouseX - slider.sliderPosX) / slider.sliderBarLength) * slider.module.getSettings().getSettingDef(slider.settingName).getMaxValue()));
                        }
                        return true;
                    }
                }
            }
        }
        return false;
    }

    public static class Slider {
        public int sliderPosX;
        public int sliderPosY;
        public int sliderBarLength;
        public double sliderKnobX;
        public Module module;
        public Setting setting;
        public String settingName;

        public Slider(int x, int y, int length, double knob, Module module, Setting setting, String settingName) {
            sliderPosX = x;
            sliderPosY = y;
            sliderBarLength = length;
            sliderKnobX = knob;
            this.module = module;
            this.setting = setting;
            this.settingName = settingName;
        }
    }
}

File: src/main/java/dev/toastmc/client/gui/click/CategoryRenderer.kt
package dev.toastmc.client.gui.click

import dev.toastmc.client.ToastClient.Companion.MODULE_MANAGER
import dev.toastmc.client.gui.TwoDRenderUtils.drawTextBox
import dev.toastmc.client.gui.TwoDRenderUtils.isMouseOverRect
import dev.toastmc.client.gui.click.components.Description
import dev.toastmc.client.gui.click.components.HiddenComponent
import dev.toastmc.client.module.Category
import dev.toastmc.client.module.Module
import dev.toastmc.client.util.ConfigUtil
import net.minecraft.client.util.math.MatrixStack
import org.lwjgl.glfw.GLFW
import java.awt.Color
import kotlin.math.roundToInt

class CategoryRenderer(
        private val clickGuiScreen: ClickGuiScreen,
        matrixStack: MatrixStack,
        mouseX: Int,
        mouseY: Int,
        val category: Category,
        clickedL: Boolean,
        clickedR: Boolean
) {
    var keybindingPressed: Boolean = false
    private var keybindingModule: Module? = null
    private var isKeyPressed: Boolean = false
    var description: Description? = null
    var hidden: HiddenComponent? = null
    private var isClickedL: Boolean
    private var isClickedR: Boolean
    var mouseX: Double
    var mouseY: Double
    private var isCategory: Boolean = true
    private fun setX(newX: Double) {
        settings.getPositions(categoryString).x = newX
    }

    private fun setY(newY: Double) {
        settings.getPositions(categoryString).y = newY
    }

    private val xInt: Int get() = settings.getPositions(categoryString).x.roundToInt()
    private val yInt: Int get() = settings.getPositions(categoryString).y.roundToInt()
    private val isMouseOverCat: Boolean
        get() {
            return isMouseOverRect(mouseX, mouseY, x - 4, y - 4, boxWidth + 4, boxHeight + 2)
        }

    fun setKeyPressed(keyPressed: Int) {
        if (keybindingPressed) {
            if (keyPressed == -1) return else if (keyPressed == GLFW.GLFW_KEY_BACKSPACE || keyPressed == GLFW.GLFW_KEY_DELETE) keybindingModule?.key = -1
            else keybindingModule?.key = keyPressed
            isKeyPressed = true
            clickGuiScreen.keybindingPressedCategory = null
            ConfigUtil.save()
        } else keybindingPressed = true
    }

    fun updateMousePos(mouseX: Double, mouseY: Double) {
        this.mouseX = mouseX
        this.mouseY = mouseY
    }

    fun updatePosition(dragX: Double, dragY: Double): Boolean {
        if (isMouseOverCat && isCategory) {
            setX(x + dragX)
            setY(y + dragY)
            settings.savePositions()
            return true
        }
        return false
    }

    val boxWidth: Int
        get() {
            return clickGuiScreen.w
        }
    val boxHeight: Int
        get() {
            return clickGuiScreen.h
        }
    val x: Double
        get() {
            return settings.getPositions(categoryString).x
        }
    val y: Double
        get() {
            return settings.getPositions(categoryString).y
        }

    fun getYIteration(iteration: Int): Double {
        return y + iteration + (boxHeight * iteration)
    }

    private val categoryString: String
        get() {
            return category.toString()
        }
    private var settings = clickGuiScreen.settings

    var colors = settings.colors

    init {
        this.mouseX = mouseX.toDouble()
        this.mouseY = mouseY.toDouble()
        isClickedL = clickedL
        isClickedR = clickedR
        var moduleTextColor: Int
        var moduleBgColor: Int
        val catExpanded = settings.getPositions(category.toString()).expanded
        var catBgColor: Int = Color(55, 175, 0, 200).rgb
        if (isMouseOverRect(mouseX.toDouble(), mouseY.toDouble(), x, y, boxWidth, boxHeight)) {
            if (isClickedR) {
                catBgColor = Color(55, 175, 0, 100).rgb
                settings.getPositions(categoryString).expanded = !settings.getPositions(categoryString).expanded
                settings.savePositions()
            } else {
                catBgColor = Color(55, 175, 0, 100).rgb
            }
        }
        drawTextBox(matrixStack, xInt, yInt, boxWidth, boxHeight, colors.categoryBoxColor, colors.categoryTextColor, colors.categoryPrefixColor, catBgColor, if (catExpanded) "- " else "+ ", category.toString())
        if (catExpanded) {
            var u = 1
            for (module: Module in MODULE_MANAGER.getModulesInCategory(category)) {
                if (module.enabled) {
                    moduleTextColor = colors.moduleOnTextColor
                    moduleBgColor = colors.moduleOffBgColor
                } else {
                    moduleTextColor = colors.moduleOffTextColor
                    moduleBgColor = colors.moduleOffBgColor
                }
                if (isMouseOverRect(mouseX.toDouble(), mouseY.toDouble(), x, getYIteration(u), boxWidth, boxHeight)) {
                    when {
                        isClickedL -> {
                            moduleBgColor = colors.moduleClickColor
                            module.toggle()
                            ConfigUtil.save()
                        }
                        isClickedR -> {
                            if (settings.getPositions(categoryString).expandedModules.contains(module.label)) {
                                settings.getPositions(categoryString).expandedModules.remove(module.label)
                                moduleBgColor = colors.moduleOnBgColor
                            } else {
                                settings.getPositions(categoryString).expandedModules.add(module.label)
                                moduleBgColor = colors.moduleOnBgColor
                            }
                            settings.savePositions()
                        }
                        else -> {
                            moduleBgColor = colors.moduleHoverBgColor
                        }
                    }
                    if(module.description.isNotEmpty()) description = Description(module.description, (x + boxWidth).roundToInt(), getYIteration(u).roundToInt(), true)
                }
                if(settings.moduleExpanded(categoryString, module)) moduleBgColor = colors.moduleExpandadBgColor
                drawTextBox(matrixStack, xInt, getYIteration(u).roundToInt(), boxWidth, boxHeight,
                        colors.moduleBoxColor, moduleTextColor,
                        colors.modulePrefixColor, moduleBgColor,
                        "", module.label)
                u++
                for (mod in settings.getPositions(categoryString).expandedModules){
                    if(mod == module.label){
                        HiddenComponent().render(matrixStack, module, this, "Hidden", colors, x, mouseX.toDouble(), mouseY.toDouble(), u, isClickedL)
                        u++
                        var keybindBgColor = colors.settingOnBgColor
                        var keybindText: String
                        if (isMouseOverRect(mouseX.toDouble(), mouseY.toDouble(), x, getYIteration(u), boxWidth, boxHeight)) {
                            keybindBgColor = colors.settingHoverBgColor
                            if (isClickedL) {
                                this.keybindingPressed = true
                                this.keybindingModule = module
                            }
                        }
                        if (keybindingModule === module) {
                            keybindBgColor = colors.settingHoverBgColor
                            keybindText = "Keybind: ..."
                            if (!isKeyPressed) {
                                keybindingPressed = true
                                keybindingModule = module
                            }
                        } else {
                            val key: Int = module.key
                            keybindText = if (key == GLFW.GLFW_KEY_UNKNOWN) "Keybind: NONE" else "Keybind: " + GLFW.glfwGetKeyName(key, GLFW.glfwGetKeyScancode(key))
                        }
                        drawTextBox(matrixStack, xInt, getYIteration(u).roundToInt(), boxWidth, boxHeight, colors.settingBoxColor, colors.settingOnTextColor, colors.settingPrefixColor, keybindBgColor, colors.settingPrefix, keybindText)
                        u++
                    }
                }
            }
        }
    }
}